registry/handlers/images.go

```diff
@@ -83,6 +91,64 @@ func (imh *imageManifestHandler) GetImageManifest(w http.ResponseWriter, r *http
 		return
 	}
 
+	supportsSchema2 := false
+	supportsManifestList := false
+	if acceptHeaders, ok := r.Header["Accept"]; ok {
+		for _, mediaType := range acceptHeaders {
+			if mediaType == schema2.MediaTypeManifest {
+				supportsSchema2 = true
+			}
+			if mediaType == manifestlist.MediaTypeManifestList {
+				supportsManifestList = true
+			}
+		}
+	}
+
+	// Only rewrite schema2 manifests when they are being fetched by tag.
+	// If they are being fetched by digest, we can't return something not
+	// matching the digest.
+	if schema2Manifest, isSchema2 := manifest.(*schema2.DeserializedManifest); imh.Tag != "" && isSchema2 && !supportsSchema2 {
```

Comment:
I can't think of a way to do it as a type switch that would be cleaner, but let me know if you have any ideas.

One thing that may make the code a little more readable is to move the `schema2Manifest, isSchema2 := manifest.(*schema2.DeserializedManifest)` and `manifestList, isManifestList := manifest.(*manifestlist.DeserializedManifestList)` out of the if/else statements, and only have the combinational logic in the if/else.